### PR TITLE
SM to broker communication (SM as a platform)

### PR DESCRIPTION
**Pull request by KirilKabakchiev**
The following PR introduces logic in the SM /v1/service_instances and /v1/service_bindings APIs to call the underlying broker when instance is created/deleted or binding is created/deleted.

The following changes/additions are made:

* all entities have ready field
* scheduler now supports delayed rescheduling
* scheduler now checks for concurrent async requests and returns proper errors
* scheduler supports scheduling of both sync and async actions
* scheduler analyses action errors and takes decision whether he has to reschedule the action for deletion (a.k.a. orphan mitigation)
* fix goroutine leak in scheduler
* scheduler properly updates both operation and ready field of entity when action finishes
* new fields "reschedule" and "deletion_scheduled" are added to operations showing if an operation can be retried or if it regular deletion for it should be attempted
* service plan bindable and updateable fields are now nullable so that we can support the scenario of unset bindable/updatable field on plan level and default to the value from the service
* service instances and bindings now return context in json response
* sm operations now return http errors (we make sure proper user facing errors are generated and internal info is not revealed)
* rest api bulk delete for instances and bindings is disabled
* two new interceptors are introduced to handle sync and async communication with actual brokers when instances are created/deleted and bindings are created/deleted. The interceptors take care to update the corresponding operation when a certain milestone is reached (the operation is now retryable (reschedule) or the instance/binding requires orphan mitigation(deletion_scheduled).

### CODE REVIEW

File: application.yml [hunk 1]
```diff
@@ -1,37 +1,42 @@
 server:
-  request_timeout: 4000ms
-  shutdown_timeout: 4000ms
-  port: 8085
+  request_timeout: 4000000ms
```
Comment (georgifarashev) [inline application.yml]: Do we need such long timeouts?

File: storage/interceptors/smaap_service_instance_interceptor.go [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+			if provisionResponse.Async {
+				log.C(ctx).Infof("Successful asynchronous provisioning request %+v to broker %s returned response %+v",
+					provisionRequest, broker.Name, provisionResponse)
+				operation.Reschedule = true
+				if provisionResponse.OperationKey != nil {
+					operation.ExternalID = string(*provisionResponse.OperationKey)
+				}
+
+				if _, err := i.repository.Update(ctx, operation, query.LabelChanges{}); err != nil {
+					return nil, fmt.Errorf("failed to update operation with id %s to mark that next execution should be a reschedule", instance.ID)
```
Comment (georgifarashev) [inline storage/interceptors/smaap_service_instance_interceptor.go]: Add the error in the log.

File: storage/interceptors/smaap_service_instance_interceptor.go [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+		case <-ticker.C:
+			log.C(ctx).Infof("Sending poll last operation request %+v for instance with id %s and name %s", pollingRequest, instance.ID, instance.Name)
+			pollingResponse, err := osbClient.PollLastOperation(pollingRequest)
+			if err != nil {
+				if osbc.IsGoneError(err) && operation.Type == types.DELETE {
+					log.C(ctx).Infof("Successfully finished polling operation for instance with id %s and name %s", instance.ID, instance.Name)
+
+					operation.Reschedule = false
+					if _, err := i.repository.Update(ctx, operation, query.LabelChanges{}); err != nil {
+						return fmt.Errorf("failed to update operation with id %s to mark that next execution should be a reschedule", operation.ID)
```
Comment (georgifarashev) [inline storage/interceptors/smaap_service_instance_interceptor.go]: Add the error in the log.

File: storage/interceptors/smaap_service_instance_interceptor.go [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+			switch pollingResponse.State {
+			case osbc.StateInProgress:
+				log.C(ctx).Infof("Polling of instance still in progress. Rescheduling polling last operation request %+v to for provisioning of instance with id %s and name %s...", pollingRequest, instance.ID, instance.Name)
+
+			case osbc.StateSucceeded:
+				log.C(ctx).Infof("Successfully finished polling operation for instance with id %s and name %s", instance.ID, instance.Name)
+
+				operation.Reschedule = false
+				if _, err := i.repository.Update(ctx, operation, query.LabelChanges{}); err != nil {
+					return fmt.Errorf("failed to update operation with id %s to mark that next execution should be a reschedule", operation.ID)
```
Comment (georgifarashev) [inline storage/interceptors/smaap_service_instance_interceptor.go]: Add the error in the log.

File: storage/interceptors/smaap_service_instance_interceptor.go [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+				operation.Reschedule = true
+				if provisionResponse.OperationKey != nil {
+					operation.ExternalID = string(*provisionResponse.OperationKey)
+				}
+
+				if _, err := i.repository.Update(ctx, operation, query.LabelChanges{}); err != nil {
+					return nil, fmt.Errorf("failed to update operation with id %s to mark that next execution should be a reschedule", instance.ID)
+				}
+			} else {
+				log.C(ctx).Infof("Successful synchronous provisioning %+v to broker %s returned response %+v",
```
Comment (dpanayotov) [inline storage/interceptors/smaap_service_instance_interceptor.go]: set usable=true here

File: storage/interceptors/smaap_service_instance_interceptor.go [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+				return nil, fmt.Errorf("failed to marshal OSB context %+v: %s", provisionRequest.Context, err)
+			}
+			instance.Context = contextBytes
+
+			log.C(ctx).Infof("Sending provision request %+v to broker with name %s", provisionRequest, broker.Name)
+			provisionResponse, err = osbClient.ProvisionInstance(provisionRequest)
+			if err != nil {
+				brokerError := &util.HTTPError{
+					ErrorType:   "BrokerError",
+					Description: fmt.Sprintf("Failed provisioning request %+v: %s", provisionRequest, err),
```
Comment (dpanayotov) [inline storage/interceptors/smaap_service_instance_interceptor.go]: request might contain parameters which might be sensitive

File: storage/interceptors/smaap_service_instance_interceptor.go [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+		var provisionResponse *osbc.ProvisionResponse
+		if !operation.Reschedule {
+			provisionRequest := i.prepareProvisionRequest(instance, service.CatalogID, plan.CatalogID)
+			contextBytes, err := json.Marshal(provisionRequest.Context)
+			if err != nil {
+				return nil, fmt.Errorf("failed to marshal OSB context %+v: %s", provisionRequest.Context, err)
+			}
+			instance.Context = contextBytes
+
+			log.C(ctx).Infof("Sending provision request %+v to broker with name %s", provisionRequest, broker.Name)
```
Comment (dpanayotov) [inline storage/interceptors/smaap_service_instance_interceptor.go]: strip parameters

File: storage/interceptors/smaap_service_binding_interceptor.go [hunk 8]
```diff
@@ -0,0 +1,521 @@
...
+					operation.DeletionScheduled = time.Now()
+					operation.Reschedule = false
+					if _, err := i.repository.Update(ctx, operation, query.LabelChanges{}); err != nil {
+						return nil, fmt.Errorf("failed to update operation with id %s to schedule orphan mitigation after broker error %s: %s", operation.ID, brokerError, err)
+					}
+				}
+				return nil, brokerError
+			}
+
+			if err := i.enrichBindingWithBindingResponse(binding, bindResponse); err != nil {
```
Comment (georgifarashev) [inline storage/interceptors/smaap_service_binding_interceptor.go]: If binding is done asynchronously the binding response does not contain the credentials. After successful operation (the poll on line 205). GET request should be done in order to get the binding details.

Comment (dpanayotov) [reply storage/interceptors/smaap_service_binding_interceptor.go]: 👍 Only if `"bindings_retrievable" :true`

Comment (KirilKabakchiev) [reply storage/interceptors/smaap_service_binding_interceptor.go]: and if false do we trigger orphan mitigation ?

Comment (KirilKabakchiev) [reply storage/interceptors/smaap_service_binding_interceptor.go]: so what im thinking - if bindings_retrievable is false, we will send accepts_incomplete=false to the broker on Bind request (this way we say that if the broker does not support get binding, we only support sync bind broker calls). If bindings_retrievable is true, we send accepts_incomplete=true and if the response is async and then our Get binding call fails, we schedule orphan mitigation. Any opinions ?

File: storage/interceptors/smaap_service_binding_interceptor.go [hunk 9]
```diff
@@ -0,0 +1,521 @@
...
+
+		if isBindable := !i.isPlanBindable(service, plan); isBindable {
+			return nil, &util.HTTPError{
+				ErrorType:   "BrokerError",
+				Description: fmt.Sprintf("plan %s is not bindable", plan.CatalogName),
+				StatusCode:  http.StatusBadGateway,
+			}
+		}
+
+		if isReady := i.isInstanceReady(instance); !isReady {
```
Comment (dpanayotov) [inline storage/interceptors/smaap_service_binding_interceptor.go]: From OSB spec:
`The Platform SHOULD NOT allow the creation of new bindings of an unusable Service Instance until the instance has been deleted or repaired by a subsequent update. `

File: storage/interceptors/smaap_service_instance_interceptor.go [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+			case osbc.StateFailed:
+				log.C(ctx).Infof("Failed polling operation for instance with id %s and name %s", instance.ID, instance.Name)
+				operation.Reschedule = false
+				if enableOrphanMitigation {
+					operation.DeletionScheduled = time.Now()
+				}
+				if _, err := i.repository.Update(ctx, operation, query.LabelChanges{}); err != nil {
+					return fmt.Errorf("failed to update operation with id %s after failed of last operation for instance with id %s", operation.ID, instance.ID)
+				}
+
```
Comment (dpanayotov) [inline storage/interceptors/smaap_service_instance_interceptor.go]: update `usable` field from error response `instance_usable` in order to forbid binding requests

Comment (KirilKabakchiev) [reply storage/interceptors/smaap_service_instance_interceptor.go]: there is no OSB spec release with this yet (2.15 does not include it)

**Pull request merged**